File: Sources/Sentry/SentryExtraContextProvider.m
#import "SentryExtraContextProvider.h"
#import "SentryCrashIntegration.h"
#import "SentryCrashWrapper.h"
#import "SentryLogC.h"
#import "SentryNSProcessInfoWrapper.h"
#import "SentryUIDeviceWrapper.h"

NSString *const kSentryProcessInfoThermalStateNominal = @"nominal";
NSString *const kSentryProcessInfoThermalStateFair = @"fair";
NSString *const kSentryProcessInfoThermalStateSerious = @"serious";
NSString *const kSentryProcessInfoThermalStateCritical = @"critical";

@interface SentryExtraContextProvider ()

@property (nonatomic, strong) SentryCrashWrapper *crashWrapper;
@property (nonatomic, strong) SentryNSProcessInfoWrapper *processInfoWrapper;

#if TARGET_OS_IOS && SENTRY_HAS_UIKIT
@property (nonatomic, strong) SentryUIDeviceWrapper *deviceWrapper;
#endif // TARGET_OS_IOS && SENTRY_HAS_UIKIT

@end

@implementation SentryExtraContextProvider

- (instancetype)initWithCrashWrapper:(id)crashWrapper
                  processInfoWrapper:(id)processInfoWrapper
#if TARGET_OS_IOS && SENTRY_HAS_UIKIT
                       deviceWrapper:(SentryUIDeviceWrapper *)deviceWrapper
#endif // TARGET_OS_IOS && SENTRY_HAS_UIKIT
{
    if (self = [super init]) {
        self.crashWrapper = crashWrapper;
        self.processInfoWrapper = processInfoWrapper;
#if TARGET_OS_IOS && SENTRY_HAS_UIKIT
        self.deviceWrapper = deviceWrapper;
#endif // TARGET_OS_IOS && SENTRY_HAS_UIKIT
    }
    return self;
}

- (NSDictionary *)getExtraContext
{
    return @{ @"device" : [self getExtraDeviceContext], @"app" : [self getExtraAppContext] };
}

- (NSDictionary *)getExtraDeviceContext
{
    NSMutableDictionary *extraDeviceContext = [[NSMutableDictionary alloc] init];

    extraDeviceContext[SentryDeviceContextFreeMemoryKey] = @(self.crashWrapper.freeMemorySize);
    extraDeviceContext[@"processor_count"] = @([self.processInfoWrapper processorCount]);

    NSProcessInfoThermalState thermalState = [self.processInfoWrapper thermalState];
    switch (thermalState) {
    case NSProcessInfoThermalStateNominal:
        extraDeviceContext[@"thermal_state"] = kSentryProcessInfoThermalStateNominal;
        break;
    case NSProcessInfoThermalStateFair:
        extraDeviceContext[@"thermal_state"] = kSentryProcessInfoThermalStateFair;
        break;
    case NSProcessInfoThermalStateSerious:
        extraDeviceContext[@"thermal_state"] = kSentryProcessInfoThermalStateSerious;
        break;
    case NSProcessInfoThermalStateCritical:
        extraDeviceContext[@"thermal_state"] = kSentryProcessInfoThermalStateCritical;
        break;
    default:
        SENTRY_LOG_WARN(@"Unexpected thermal state enum value: %ld", (long)thermalState);
        break;
    }

#if TARGET_OS_IOS && SENTRY_HAS_UIKIT
    if (self.deviceWrapper.orientation != UIDeviceOrientationUnknown) {
        extraDeviceContext[@"orientation"]
            = UIDeviceOrientationIsPortrait(self.deviceWrapper.orientation) ? @"portrait"
                                                                            : @"landscape";
    }

    if (self.deviceWrapper.isBatteryMonitoringEnabled) {
        extraDeviceContext[@"charging"]
            = self.deviceWrapper.batteryState == UIDeviceBatteryStateCharging ? @(YES) : @(NO);
        extraDeviceContext[@"battery_level"] = @((int)(self.deviceWrapper.batteryLevel * 100));
    }
#endif // TARGET_OS_IOS && SENTRY_HAS_UIKIT
    return extraDeviceContext;
}

- (NSDictionary *)getExtraAppContext
{
    NSMutableDictionary *extraAppContext = [NSMutableDictionary dictionary];
    extraAppContext[SentryDeviceContextAppMemoryKey] = @(self.crashWrapper.appMemorySize);

    if (@available(iOS 14.0, macOS 11.0, watchOS 7.0, tvOS 14.0, *)) {
        extraAppContext[@"ios_app_on_mac"] = @(self.processInfoWrapper.isiOSAppOnMac);
    }

    if (@available(iOS 13.0, macOS 10.15, watchOS 6.0, tvOS 13.0, *)) {
        extraAppContext[@"mac_catalyst_app"] = @(self.processInfoWrapper.isMacCatalystApp);
    }

    return extraAppContext;
}

@end

File: Sources/Swift/Helper/SentryExtraContextProvider.swift
// swiftlint:disable missing_docs
internal import _SentryPrivate

@_spi(Private) @objc public final class SentryExtraContextProvider: NSObject {
    
    private static let kSentryProcessInfoThermalStateNominal = "nominal"
    private static let kSentryProcessInfoThermalStateFair = "fair"
    private static let kSentryProcessInfoThermalStateSerious = "serious"
    private static let kSentryProcessInfoThermalStateCritical = "critical"
    
    private let memoryMetricsProvider: SentryMemoryMetricsProvider
    private let processInfoWrapper: SentryProcessInfoSource
    private let reachability: SentryReachability
    
    #if (os(iOS)) && !SENTRY_NO_UI_FRAMEWORK
    private let deviceWrapper: SentryUIDeviceWrapper

    init(memoryMetricsProvider: SentryMemoryMetricsProvider, processInfoWrapper: SentryProcessInfoSource, deviceWrapper: SentryUIDeviceWrapper, reachability: SentryReachability) {
        self.memoryMetricsProvider = memoryMetricsProvider
        self.processInfoWrapper = processInfoWrapper
        self.deviceWrapper = deviceWrapper
        self.reachability = reachability
    }
    #else
    init(memoryMetricsProvider: SentryMemoryMetricsProvider, processInfoWrapper: SentryProcessInfoSource, reachability: SentryReachability) {
        self.memoryMetricsProvider = memoryMetricsProvider
        self.processInfoWrapper = processInfoWrapper
        self.reachability = reachability
    }
    #endif
    
    @objc public func getExtraContext() -> [String: Any] {
        [
            "device": getExtraDeviceContext(),
            "app": getExtraAppContext()
        ]
    }
    
    private func getExtraDeviceContext() -> [String: Any] {
        var extraDeviceContext: [String: Any] = [
            SentryDeviceContextFreeMemoryKey: NSNumber(value: memoryMetricsProvider.freeMemorySize),
            "processor_count": NSNumber(value: processInfoWrapper.processorCount)
        ]

        let thermalState = processInfoWrapper.thermalState
        switch thermalState {
        case .nominal:
            extraDeviceContext["thermal_state"] = Self.kSentryProcessInfoThermalStateNominal
        case .fair:
            extraDeviceContext["thermal_state"] = Self.kSentryProcessInfoThermalStateFair
        case .serious:
            extraDeviceContext["thermal_state"] = Self.kSentryProcessInfoThermalStateSerious
        case .critical:
            extraDeviceContext["thermal_state"] = Self.kSentryProcessInfoThermalStateCritical
        default:
            SentrySDKLog.warning("Unexpected thermal state enum value: \(thermalState)")
        }

        extraDeviceContext["low_power_mode"] = NSNumber(value: processInfoWrapper.isLowPowerModeEnabled)

        // The connection type is only known while the SDK monitors connectivity, which it does as
        // long as it has a transport. `connection_type` is the device context alias of
        // `network.connection.type`, so `connection_effective_type` mirrors
        // `network.connection.effective_type` the same way.
        // https://getsentry.github.io/sentry-conventions/attributes/network/
        if let connection = reachability.currentConnection {
            extraDeviceContext["connection_type"] = connection.type
            if let effectiveType = connection.effectiveType {
                extraDeviceContext["connection_effective_type"] = effectiveType
            }
        }
        
        #if (os(iOS)) && !SENTRY_NO_UI_FRAMEWORK
        if deviceWrapper.orientation != .unknown {
            extraDeviceContext["orientation"]
            = deviceWrapper.orientation.isPortrait ? "portrait" : "landscape"
        }

        if deviceWrapper.isBatteryMonitoringEnabled {
            extraDeviceContext["charging"] = NSNumber(value: deviceWrapper.batteryState == .charging)
            extraDeviceContext["battery_level"] = NSNumber(value: Int(deviceWrapper.batteryLevel * 100))
        }
        #endif
        return extraDeviceContext
    }
    
    private func getExtraAppContext() -> [String: Any] {
        [ SentryDeviceContextAppMemoryKey: NSNumber(value: self.memoryMetricsProvider.appMemorySize) ]
    }
}
// swiftlint:enable missing_docs
